libcudacxx/include/cuda/__floating_point/cast.h

```diff
@@ -0,0 +1,486 @@
+#elif defined(_CCCL_IMPLICIT_SYSTEM_HEADER_CLANG)
+#  pragma clang system_header
+#elif defined(_CCCL_IMPLICIT_SYSTEM_HEADER_MSVC)
+#  pragma system_header
+#endif // no system header
+
+#if _CCCL_STD_VER >= 2017
+
+#  include <cuda/__floating_point/fp.h>
+#  include <cuda/__floating_point/type_traits.h>
+#  include <cuda/std/__concepts/concept_macros.h>
+#  include <cuda/std/__type_traits/is_integral.h>
+#  include <cuda/std/__type_traits/is_same.h>
+#  include <cuda/std/__type_traits/is_signed.h>
+#  include <cuda/std/__type_traits/is_unsigned.h>
+
+_LIBCUDACXX_BEGIN_NAMESPACE_CUDA
+
+struct __fp_cast
+{
+  template <class _Tp, class _Up>
+  _CCCL_NODISCARD _LIBCUDACXX_HIDE_FROM_ABI static constexpr _Tp __cast_generic(const _Up& __src)
+  {
+    _CCCL_ASSERT(false, "Unsupported floating point cast");
+    // todo: implement generic cast
+    //  - fp -> fp
+    //  - fp -> integral
+    //  - integral -> fp
+
+    _Tp __dst{};
+
+    // Copy sign
+    // if constexpr (_Tp::__is_signed && _Up::__is_signed)
+    // {
+    //   __dst.__set_sign(__src.__get_sign());
+    // }
+    // else if constexpr (!_Tp::__is_signed && _Up::__is_signed)
+    // {
+    //   if (__src.__get_sign())
+    //   {
+    //     return _Tp::__nan();
+    //   }
+    // }
+
+    // using _Sp =
+    //   _CUDA_VSTD::make_signed_t<_CUDA_VSTD::common_type_t<typename _Tp::__storage_type, typename
+    //   _Up::__storage_type>>;
+
+    // // Convert exponent
+    // constexpr _Sp __src_exp_bias = _Up::__exp_val_mask() / 2;
+    // constexpr _Sp __dst_exp_bias = _Tp::__exp_val_mask() / 2;
+
+    // _Sp __dst_exp = static_cast<_Sp>(__src.__get_exp()) - __src_exp_bias + __dst_exp_bias;
+
+    // if (__dst_exp >= static_cast<_Sp>(_Tp::__exp_val_mask()))
+    // {
+    //   return _Tp::__inf();
+    // }
+
+    // __dst.__set_exp(static_cast<typename _Tp::__storage_type>(__dst_exp));
+
+    // // Convert mantissa (todo: implement rounding)
+    // constexpr ptrdiff_t __mant_diff =
+    //   static_cast<ptrdiff_t>(_Tp::__mant_nbits) - static_cast<ptrdiff_t>(_Up::__mant_nbits);
+
+    // _Sp __dst_mant{};
+
+    // if constexpr (__mant_diff < 0)
+    // {
+    //   __dst_mant = static_cast<_Sp>(__src.__get_mant()) >> (-__mant_diff);
+    // }
+    // else
+    // {
+    //   __dst_mant = static_cast<_Sp>(__src.__get_mant()) << __mant_diff;
+    // }
+
+    // __dst.__set_mant(static_cast<typename _Tp::__storage_type>(__dst_mant));
+
+    return __dst;
+  }
+
+  _CCCL_TEMPLATE(class _Tp, class _Up)
+  _CCCL_REQUIRES(_CUDA_VSTD::is_same_v<__fp_make_config_from_t<_Up>, __fp16_config>)
+  _CCCL_NODISCARD _CCCL_HIDE_FROM_ABI _CCCL_DEVICE static _Tp __cast_impl_device(const _Up& __val) noexcept
+  {
+    fp16 __fp_val{__fp_from_native, __val};
+    _Tp __ret{};
+
+    if constexpr (_CUDA_VSTD::is_same_v<_Tp, fp32>)
+    {
+      NV_IF_TARGET(NV_IS_DEVICE,
+                   (asm("cvt.f32.f16 %0, %1;" : "=r"(__ret.__storage_) : "h"(__fp_val.__storage_)); return __ret;))
```

Comment:
We really need to move these (and that applies to basically all of the inline asms, which should also be volatile) into a codegen headers. For the beginning of me doing that (though for arithmetic and not conversions, but also including some cmake cleanup related to codegens), please see https://github.com/griwes/cccl/commit/c81edb4fba39e72f3eddfba23b6fd48d6c1d6151 in libcudacxx/codegen.